File: src/main/java/org/example/kalkulationsprogramm/service/AusgangsGeschaeftsDokumentAuditService.java
package org.example.kalkulationsprogramm.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokument;
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentAudit;
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentAuditAktion;
import org.example.kalkulationsprogramm.domain.FrontendUserProfile;
import org.example.kalkulationsprogramm.repository.AusgangsGeschaeftsDokumentAuditRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * GoBD-konforme Audit-Protokollierung für AusgangsGeschaeftsDokumente.
 *
 * <p>Zentrale Stelle für: Erstellen, Ändern, Buchen, Versenden, Stornieren, Löschen.
 * Bei Löschung und Änderung ist eine Begründung Pflicht.</p>
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AusgangsGeschaeftsDokumentAuditService {

    private final AusgangsGeschaeftsDokumentAuditRepository auditRepository;

    @Transactional
    public void protokolliereErstellung(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter, String ipAdresse) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.ERSTELLT, bearbeiter, "Initiale Erstellung", ipAdresse);
    }

    @Transactional
    public void protokolliereAenderung(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter,
                                       String aenderungsgrund, String ipAdresse) {
        if (aenderungsgrund == null || aenderungsgrund.isBlank()) {
            throw new IllegalArgumentException("Änderungsgrund ist Pflicht (GoBD)");
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GEAENDERT, bearbeiter, aenderungsgrund, ipAdresse);
    }

    @Transactional
    public void protokolliereBuchung(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter, String ipAdresse) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GEBUCHT, bearbeiter, "Festschreibung/Buchung", ipAdresse);
    }

    @Transactional
    public void protokolliereVersand(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter, String ipAdresse) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.VERSENDET, bearbeiter, "Versand an Kunden", ipAdresse);
    }

    @Transactional
    public void protokolliereStornierung(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter,
                                         String grund, String ipAdresse) {
        if (grund == null || grund.isBlank()) {
            throw new IllegalArgumentException("Stornierungsgrund ist Pflicht");
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.STORNIERT, bearbeiter, grund, ipAdresse);
    }

    /**
     * Protokolliert die Löschung eines Entwurfs. MUSS vor dem Hard-Delete aufgerufen werden,
     * damit der Snapshot noch vollständig ist. Begründung ist Pflicht (GoBD).
     */
    @Transactional
    public void protokolliereLoeschung(AusgangsGeschaeftsDokument dokument, FrontendUserProfile bearbeiter,
                                       String begruendung, String ipAdresse) {
        if (begruendung == null || begruendung.isBlank()) {
            throw new IllegalArgumentException("Begründung für Löschung ist Pflicht (GoBD)");
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GELOESCHT, bearbeiter, begruendung, ipAdresse);
    }

    @Transactional
    public void protokolliereDigitaleAnnahme(AusgangsGeschaeftsDokument dokument, String ipAdresse) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.DIGITAL_ANGENOMMEN, null,
                "Digitale Annahme durch Kunden", ipAdresse);
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> getHistorie(Long dokumentId) {
        return auditRepository.findByDokumentIdOrderByGeaendertAmDesc(dokumentId)
                .stream()
                .map(this::toMap)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> getHistorieByNummer(String dokumentNummer) {
        return auditRepository.findByDokumentNummerOrderByGeaendertAmDesc(dokumentNummer)
                .stream()
                .map(this::toMap)
                .collect(Collectors.toList());
    }

    private void save(AusgangsGeschaeftsDokument dokument, AusgangsGeschaeftsDokumentAuditAktion aktion,
                      FrontendUserProfile bearbeiter, String grund, String ipAdresse) {
        AusgangsGeschaeftsDokumentAudit audit = AusgangsGeschaeftsDokumentAudit.fromDokument(
                dokument, aktion, bearbeiter, grund, ipAdresse);
        auditRepository.save(audit);
        log.info("Audit-Eintrag: {} für Dokument {} (Nr: {}) durch {} – {}",
                aktion, dokument.getId(), dokument.getDokumentNummer(),
                bearbeiter != null ? bearbeiter.getId() : "system", grund);
    }

    private Map<String, Object> toMap(AusgangsGeschaeftsDokumentAudit a) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("id", a.getId());
        map.put("aktion", a.getAktion().name());
        map.put("dokumentId", a.getDokumentId());
        map.put("dokumentNummer", a.getDokumentNummer());
        map.put("typ", a.getTyp().name());
        map.put("betragNetto", a.getBetragNetto());
        map.put("betragBrutto", a.getBetragBrutto());
        map.put("gebucht", a.isGebucht());
        map.put("storniert", a.isStorniert());
        map.put("digitalAngenommen", a.isDigitalAngenommen());
        map.put("inhaltHash", a.getInhaltHash());
        map.put("geaendertVon", a.getGeaendertVon() != null ? a.getGeaendertVon().getDisplayName() : null);
        map.put("geaendertVonId", a.getGeaendertVon() != null ? a.getGeaendertVon().getId() : null);
        map.put("geaendertAm", a.getGeaendertAm() != null ? a.getGeaendertAm().toString() : null);
        map.put("aenderungsgrund", a.getAenderungsgrund());
        map.put("ipAdresse", a.getIpAdresse());
        return map;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/AusgangsGeschaeftsDokumentAuditService.kt
package org.example.kalkulationsprogramm.service

import jakarta.persistence.EntityManager
import org.example.kalkulationsprogramm.domain.AuditChainState
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokument
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentAudit
import org.example.kalkulationsprogramm.domain.AusgangsGeschaeftsDokumentAuditAktion
import org.example.kalkulationsprogramm.domain.FrontendUserProfile
import org.example.kalkulationsprogramm.repository.AuditChainStateRepository
import org.example.kalkulationsprogramm.repository.AusgangsGeschaeftsDokumentAuditRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.LinkedHashMap

@Service
class AusgangsGeschaeftsDokumentAuditService(
    private val auditRepository: AusgangsGeschaeftsDokumentAuditRepository,
    private val chainStateRepository: AuditChainStateRepository,
    private val entityManager: EntityManager,
) {
    @Transactional
    fun protokolliereErstellung(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        ipAdresse: String?,
    ) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.ERSTELLT, bearbeiter, "Initiale Erstellung", ipAdresse)
    }

    @Transactional
    fun protokolliereAenderung(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        aenderungsgrund: String?,
        ipAdresse: String?,
    ) {
        if (aenderungsgrund.isNullOrBlank()) {
            throw IllegalArgumentException("Änderungsgrund ist Pflicht (GoBD)")
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GEAENDERT, bearbeiter, aenderungsgrund, ipAdresse)
    }

    @Transactional
    fun protokolliereBuchung(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        ipAdresse: String?,
    ) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GEBUCHT, bearbeiter, "Festschreibung/Buchung", ipAdresse)
    }

    @Transactional
    fun protokolliereVersand(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        ipAdresse: String?,
    ) {
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.VERSENDET, bearbeiter, "Versand an Kunden", ipAdresse)
    }

    @Transactional
    fun protokolliereStornierung(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        grund: String?,
        ipAdresse: String?,
    ) {
        if (grund.isNullOrBlank()) {
            throw IllegalArgumentException("Stornierungsgrund ist Pflicht")
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.STORNIERT, bearbeiter, grund, ipAdresse)
    }

    @Transactional
    fun protokolliereLoeschung(
        dokument: AusgangsGeschaeftsDokument,
        bearbeiter: FrontendUserProfile?,
        begruendung: String?,
        ipAdresse: String?,
    ) {
        if (begruendung.isNullOrBlank()) {
            throw IllegalArgumentException("Begründung für Löschung ist Pflicht (GoBD)")
        }
        save(dokument, AusgangsGeschaeftsDokumentAuditAktion.GELOESCHT, bearbeiter, begruendung, ipAdresse)
    }

    @Transactional
    fun protokolliereDigitaleAnnahme(dokument: AusgangsGeschaeftsDokument, ipAdresse: String?) {
        save(
            dokument,
            AusgangsGeschaeftsDokumentAuditAktion.DIGITAL_ANGENOMMEN,
            null,
            "Digitale Annahme durch Kunden",
            ipAdresse,
        )
    }

    @Transactional(readOnly = true)
    fun getHistorie(dokumentId: Long?): List<Map<String, Any?>> =
        auditRepository.findByDokumentIdOrderByGeaendertAmDesc(dokumentId).map(::toMap)

    @Transactional(readOnly = true)
    fun getHistorieByNummer(dokumentNummer: String): List<Map<String, Any?>> =
        auditRepository.findByDokumentNummerOrderByGeaendertAmDesc(dokumentNummer).map(::toMap)

    private fun save(
        dokument: AusgangsGeschaeftsDokument,
        aktion: AusgangsGeschaeftsDokumentAuditAktion,
        bearbeiter: FrontendUserProfile?,
        grund: String?,
        ipAdresse: String?,
    ) {
        val audit = AusgangsGeschaeftsDokumentAudit.fromDokument(dokument, aktion, bearbeiter, grund, ipAdresse)
        appendToChain(audit)
        log.info(
            "Audit-Eintrag #{} (chain={}): {} für Dokument {} (Nr: {}) durch {} – {}",
            audit.id,
            audit.chainIndex,
            aktion,
            dokument.id,
            dokument.dokumentNummer,
            bearbeiter?.id ?: "system",
            grund,
        )
    }

    @Transactional(propagation = Propagation.REQUIRED)
    fun appendToChain(audit: AusgangsGeschaeftsDokumentAudit): AusgangsGeschaeftsDokumentAudit {
        var state = chainStateRepository.lockState()
        if (state == null) {
            state = AuditChainState()
            state.id = 1
            state.lastChainIndex = -1L
            state.lastEntryHash = null
            state.updatedAt = LocalDateTime.now()
            state = chainStateRepository.saveAndFlush(state)
        }

        val nextIndex = (state.lastChainIndex ?: -1L) + 1
        audit.chainIndex = nextIndex
        audit.previousHash = state.lastEntryHash

        val saved = auditRepository.saveAndFlush(audit)
        entityManager.refresh(saved)
        saved.entryHash = saved.computeEntryHash()
        auditRepository.saveAndFlush(saved)

        state.lastChainIndex = nextIndex
        state.lastEntryHash = saved.entryHash
        state.updatedAt = LocalDateTime.now()
        chainStateRepository.saveAndFlush(state)

        return saved
    }

    private fun toMap(audit: AusgangsGeschaeftsDokumentAudit): Map<String, Any?> {
        val map = LinkedHashMap<String, Any?>()
        map["id"] = audit.id
        map["chainIndex"] = audit.chainIndex
        map["aktion"] = audit.aktion?.name
        map["dokumentId"] = audit.dokumentId
        map["dokumentNummer"] = audit.dokumentNummer
        map["typ"] = audit.typ?.name
        map["betragNetto"] = audit.betragNetto
        map["betragBrutto"] = audit.betragBrutto
        map["gebucht"] = audit.isGebucht()
        map["storniert"] = audit.isStorniert()
        map["digitalAngenommen"] = audit.isDigitalAngenommen()
        map["inhaltHash"] = audit.inhaltHash
        map["previousHash"] = audit.previousHash
        map["entryHash"] = audit.entryHash
        map["geaendertVon"] = audit.geaendertVon?.displayName
        map["geaendertVonId"] = audit.geaendertVon?.id
        map["geaendertAm"] = audit.geaendertAm?.toString()
        map["aenderungsgrund"] = audit.aenderungsgrund
        map["ipAdresse"] = audit.ipAdresse
        return map
    }

    companion object {
        private val log = LoggerFactory.getLogger(AusgangsGeschaeftsDokumentAuditService::class.java)
    }
}
